File: lua/SledgehammerLua/media/lua/client/Sledgehammer/Modules/Factions/Objects/Faction.lua
-- This file is part of Sledgehammer.
--
--    Sledgehammer is free software: you can redistribute it and/or modify
--    it under the terms of the GNU Lesser General Public License as published by
--    the Free Software Foundation, either version 3 of the License, or
--    (at your option) any later version.
--
--    Sledgehammer is distributed in the hope that it will be useful,
--    but WITHOUT ANY WARRANTY; without even the implied warranty of
--    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
--    GNU Lesser General Public License for more details.
--
--    You should have received a copy of the GNU Lesser General Public License
--    along with Sledgehammer. If not, see <http://www.gnu.org/licenses/>.
--
--	Sledgehammer is free to use and modify, ONLY for non-official third-party servers 
--    not affiliated with TheIndieStone, or it's immediate affiliates, or contractors. 

require "Sledgehammer"
require "Sledgehammer/Objects/Color"
require "Sledgehammer/Modules/Factions/Objects/FactionMember"
require "Sledgehammer/Modules/Factions/Objects/FactionRelationship"

----------------------------------------------------------------
-- Faction.Lua
-- 
-- @module Factions
-- @author Jab
-- @license LGPL
--
-- @string name 	Name of the Faction.
-- @string tag 		The Tag for the Faction in chat.
-- @Color color 	The color of the Faction.
----------------------------------------------------------------
Faction = class(function(o, name, tag, color)
	
	-- Set the variables.
	o.name          = name;
	o.tag           = tag;
	o.color         = color;
	o.members       = {};
	o.relationships = {};
end);

----------------------------------------------------------------
-- @int id 	The internal ID of the Player.
-- @return Returns the Member of the Faction associated with the id.
----------------------------------------------------------------
function Faction:getMember(id)
	member = nil;
	if id ~= -1 then
		local length = tLength(self.members) - 1;
		for index=0, length, 1 do
			nextMember = self.members[index];
			if nextMember.getID() == id then
				member = nextMember;
				break
			end
		end
	end
	return member;
end

----------------------------------------------------------------
-- @number id 	The ID of the other Faction.
-- @return 		Returns the Relationship associated with the Faction ID given.
----------------------------------------------------------------
function Faction:getRelationshipFrom(id)
	relationship = nil;
	if id ~= -1 then
		local length = tLength(self.relationships) - 1;
		for index=0, length, 1 do
			nextRelationship = self.relationships[index];
			if nextRelationship.getIDFrom() == id then
				relationship = nextRelationship;
				break
			end
		end
	end
	return relationship;
end

----------------------------------------------------------------
-- @number id 	The ID of the other Faction.
-- @return 		Returns the Relationship associated with the Faction ID given.
----------------------------------------------------------------
function Faction:getRelationshipTo(ID_to)
	relationship = nil;
	if id ~= -1 then
		local length = tLength(self.relationships) - 1;
		for index=0, length, 1 do
			nextRelationship = self.relationships[index];
			if nextRelationship.getIDTo() == ID_to then
				relationship = nextRelationship;
				break
			end
		end
	end
	return relationship;
end

----------------------------------------------------------------
-- @return 	Returns the ID of the Faction.
----------------------------------------------------------------
function Faction:getID()
	return self.ID;
end

----------------------------------------------------------------
-- @private
-- @number id 	The ID of the Faction.
----------------------------------------------------------------
function Faction:setID(id)
	self.ID = id;
end

----------------------------------------------------------------
-- @return 	Returns the name of the Faction.
----------------------------------------------------------------
function Faction:getName()
	return self.name;
end

----------------------------------------------------------------
-- @return 	Returns the Tag of the Faction for chat.
----------------------------------------------------------------
function Faction:getTag()
	return self.tag;
end

----------------------------------------------------------------
-- @return Returns the Color of the Faction.
----------------------------------------------------------------
function Faction:getColor()
	return self.color;
end

----------------------------------------------------------------
-- @return Returns the list of FactionMembers.
----------------------------------------------------------------
function Faction:getMembers()
	return self.members;
end

----------------------------------------------------------------
-- @protected
-- @table members 	List of Members to be set for the Faction.
----------------------------------------------------------------
function Faction:setMembers(members)
	self.members = members;
end

----------------------------------------------------------------
-- @return 	Returns the list of Relationships associated with this Faction.
----------------------------------------------------------------
function Faction:getRelationships()
	return self.relationships;
end

----------------------------------------------------------------
-- @private
-- @table relationships 	list of Relationships to be set for the Faction.
----------------------------------------------------------------
function Faction:setRelationships(relationships)
	self.relationships = relationships;
end

File: src/sledgehammer/lua/faction/Faction.java
/*
This file is part of Sledgehammer.

   Sledgehammer is free software: you can redistribute it and/or modify
   it under the terms of the GNU Lesser General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   Sledgehammer is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License
   along with Sledgehammer. If not, see <http://www.gnu.org/licenses/>.
 */
package sledgehammer.lua.faction;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import se.krka.kahlua.vm.KahluaTable;
import sledgehammer.SledgeHammer;
import sledgehammer.database.module.faction.MongoFaction;
import sledgehammer.lua.LuaTable;
import sledgehammer.lua.chat.ChatChannel;
import sledgehammer.lua.core.Color;
import sledgehammer.lua.core.Player;
import sledgehammer.util.StringUtils;

/**
 * TODO: Document
 * 
 * @author Jab
 */
public class Faction extends LuaTable {

	private MongoFaction mongoFaction;

	private List<FactionMember> listMembers;

	private ChatChannel chatChannel;

	public Faction(MongoFaction mongoFaction) {
		super("Faction");
		setMongoDocument(mongoFaction);
		listMembers = new ArrayList<>();
	}

	public Faction(KahluaTable table) {
		super("Faction", table);
	}

	@Override
	public void onLoad(KahluaTable table) {
		// TODO: Implement.
	}

	@Override
	public void onExport() {
		// TODO: Implement.
	}

	/**
	 * TODO: Document
	 * 
	 * @param factionMember
	 * @return
	 */
	public boolean addMember(FactionMember factionMember) {
		boolean returned = false;
		if (!listMembers.contains(factionMember)) {
			listMembers.add(factionMember);
			returned = true;
			// Add the Player to the ChatChannel if he is online.
			Player player = SledgeHammer.instance.getPlayer(factionMember.getPlayerId());
			if (player != null) {
				getChatChannel().addPlayer(player, true);
			} else {
				errorln("addMember() -> Player is null.");
			}
		}
		return returned;
	}

	/**
	 * Removes a <FactionMember> from the <Faction>.
	 * 
	 * @param factionMember
	 *            The <FactionMember> member being removed.
	 */
	public void removeMember(FactionMember factionMember) {
		this.listMembers.remove(factionMember);
		// Remove the Player from the ChatChannel if he is online.
		Player player = SledgeHammer.instance.getPlayer(factionMember.getPlayerId());
		if (player != null) {
			getChatChannel().removePlayer(player);
		}
	}

	@Override
	public boolean equals(Object other) {
		if (other instanceof Faction) {
			return ((Faction) other).getUniqueId().equals(getUniqueId());
		}
		return false;
	}

	/**
	 * @param password
	 *            The <String> password to test.
	 * @return Returns whether or not the provided <String> password matches the
	 *         encrypted password.
	 */
	public boolean isPassword(String password) {
		// Grab the encrypted password for the faction.
		String passwordActualEncrypted = getEncryptedPassword();
		// If a password and one is not provided.
		if (password == null || password.isEmpty()) {
			// If the actual password is not defined, this situation is valid.
			if (passwordActualEncrypted == null || passwordActualEncrypted.isEmpty()) {
				return true;
			}
			// If the faction has a password and one isn't provided, this is invalid.
			else {
				return false;
			}
		}
		// If the actual password is not defined.
		if (passwordActualEncrypted == null || passwordActualEncrypted.isEmpty()) {
			// If the password given is defined, this situation is invalid.
			if (password != null && !password.isEmpty()) {
				return false;
			}
		}
		// Encrypt the new password.
		String passwordEncrypted = StringUtils.md5(password);
		// Test if the both encrypted passwords match.
		return passwordActualEncrypted.equals(passwordEncrypted);
	}

	public void setFactionName(String factionName, boolean save) {
		getMongoDocument().setFactionName(factionName, save);
		ChatChannel chatChannel = getChatChannel();
		if (chatChannel != null) {
			chatChannel.rename("Faction_" + factionName, true);
		}
	}

	public void setFactionTag(String tag, boolean save) {
		getMongoDocument().setFactionTag(tag, save);
		for (FactionMember factionMember : listMembers) {
			Player player = SledgeHammer.instance.getPlayer(factionMember.getPlayerId());
			if (player != null) {
				player.setNickname("[" + getFactionTag() + "] " + player.getUsername());
			}
		}
	}

	public void setFactionColor(String colorNew, boolean save) {
		getMongoDocument().setFactionColorString(colorNew, save);
		Color color = Color.getColor(colorNew);
		for (FactionMember factionMember : listMembers) {
			Player player = SledgeHammer.instance.getPlayer(factionMember.getPlayerId());
			if (player != null) {
				player.setColor(color);
			}
		}
	}

	/**
	 * Sets the new <String> password for the <Faction>.
	 * 
	 * @param password
	 *            The <String> password to set.
	 * @param save
	 *            Whether or not to save the changes.
	 */
	public void setPassword(String password, boolean save) {
		getMongoDocument().setPassword(password, save);
	}

	private String getEncryptedPassword() {
		return getMongoDocument().getEncryptedPassword();
	}

	/**
	 * Returns whether or not the <String> tag given is the same tag as the one the
	 * <Faction> uses.
	 * 
	 * @param tag
	 *            The <String> tag being tested.
	 * @return Returns true if the tag is the same as the tag being used by the
	 *         <Faction>.
	 */
	public boolean isTag(String tag) {
		return getFactionTag().toUpperCase().equals(tag.toUpperCase());
	}

	/**
	 * @return Returns the <ChatChannel> representing the <Faction>.
	 */
	public ChatChannel getChatChannel() {
		return this.chatChannel;
	}

	/**
	 * Sets the <ChatChannel> representing the <Faction>
	 * 
	 * @param chatChannel
	 *            The <ChatChannel> to set.
	 */
	public void setChatChannel(ChatChannel chatChannel) {
		this.chatChannel = chatChannel;
	}

	public String getFactionName() {
		return getMongoDocument().getFactionName();
	}

	public UUID getUniqueId() {
		return getMongoDocument().getUniqueId();
	}

	public void setUniqueId(UUID uniqueId, boolean save) {
		getMongoDocument().setUniqueId(uniqueId, save);
	}

	public MongoFaction getMongoDocument() {
		return this.mongoFaction;
	}

	public void setMongoDocument(MongoFaction mongoFaction) {
		this.mongoFaction = mongoFaction;
	}

	public UUID getOwnerId() {
		return getMongoDocument().getOwnerId();
	}

	public boolean isOwner(Player player) {
		return isOwner(player.getUniqueId());
	}

	public boolean isOwner(FactionMember factionMember) {
		return isOwner(factionMember.getPlayerId());
	}

	public boolean isOwner(UUID playerId) {
		return playerId.equals(getOwnerId());
	}

	public String getFactionTag() {
		return getMongoDocument().getFactionTag();
	}

	public String getFactionColor() {
		return getMongoDocument().getFactionColor();
	}

	public List<FactionMember> getMembers() {
		return this.listMembers;
	}

	public void setOwner(FactionMember factionMember, boolean save) {
		getMongoDocument().setOwnerId(factionMember.getPlayerId(), save);
	}

	public String getFactionRawColor() {
		return getMongoDocument().getFactionRawColor();
	}
}